waspc/src/Analyzer/TypeDefinitions/Type.hs

```diff
@@ -0,0 +1,38 @@
+module Analyzer.TypeDefinitions.Type
+  ( EnumType (..),
+    DeclType (..),
+    TypeDefinitions (..),
+  )
+where
+
+import Analyzer.Evaluator.Decl (Decl)
+import Analyzer.Evaluator.EvaluationError (EvaluationError)
+import Analyzer.Type (Type)
+import Analyzer.TypeChecker.AST (TypedExpr)
+import qualified Data.HashMap.Strict as M
+
+data EnumType = EnumType
+  { etName :: String,
+    etVariants :: [String]
+  }
+
+data DeclType = DeclType
```

Comment:
I guess `DeclType` is a value-level representation of `IsDeclType`. They are equivalent, except the value-level representation is a bit more powerful, since it is a value. It would be possible to remove `IsDeclType` and `IsEnumType` completely, but using the type class is debatably more convenient.

It is not intended to use a `DeclType` without an `IsDeclType` instance. Since the insides of `TypeDefinitions` is exported (for testing), it is possible to break it, but if that is changed/fixed, then the only way to add decl types is with `addDeclType`, which requires the class instance.

(for some more info about value-level "type classes", read https://www.haskellforall.com/2012/05/scrap-your-type-classes.html, although I don't really agree with the conclusions)

Reply:
> I guess DeclType is a value-level representation of IsDeclType.

Yes, I also see it this way.

I am asking these questions because I feel there is duplication between these instances and types, and it seems to me it is introducing and additional level of complexity, so I would like to find a way to somehow simplify it.

Few more questions then:
1. Could we go with just `DeclType` and `EnumType`, without the `IsDeclType` and `IsEnumType`? You said yes, but that with classes it is debatably more convenient -> how would that debate go, what do classes bring to the table in our case?
2. Could we go with just `IsDeclType` and `IsEnumType`, without the `DeclType` and `EnumType`? It seems to me like we couldn't, since we need to somehow "enlist" them as standard type definitions, and I don't see how can we reference instances from the value-level code.

Thanks for the article, I skimmed it and I plan to read it in more depth in following days. I was reading about a year ago about `data` vs `class` regarding dependency injection, and similar claims were being made, that `data` (explicit dependency injection) has benefits over `class` (implicit dependency injection). In our case here, I think we are doing dependency injection, so this is relevant.

However, all these discussions talk about data vs class -> instead, we have both! I guess what I would like is to have only one of them, to avoid duplication and complexity. And it seems to me like data is playing an important role, while class is maybe just a left-over that is not bringing much anymore. But let's figure this out through discussion, hopefully answers to my questions above will help.

Reply:
Ok, I went with the PR!